src/utils/buffers/histogram_buffer.rs

```diff
@@ -0,0 +1,367 @@
+    /// sample rate like so:
+    ///
+    /// ```
+    /// fn initialize(
+    ///     &mut self,
+    ///     _audio_io_layout: &AudioIOLayout,
+    ///     buffer_config: &BufferConfig,
+    ///     _context: &mut impl InitContext<Self>,
+    /// ) -> bool {
+    ///     match self.histogram_buffer.lock() {
+    ///         Ok(mut buffer) => {
+    ///             buffer.set_sample_rate(buffer_config.sample_rate);
+    ///         }
+    ///         Err(_) => return false,
+    ///     }
+    ///
+    ///     true
+    /// }
+    /// ```
+    pub fn set_sample_rate(self: &mut Self, sample_rate: f32) {
+        self.sample_rate = sample_rate;
+        self.update();
+        self.clear();
+    }
+
+    fn decay_weight(decay: f32, sample_rate: f32) -> f32 {
+        0.25f64.powf((decay as f64 * sample_rate as f64).recip()) as f32
+    }
+
+    fn update(self: &mut Self) {
+        // calculate the linear edge values from MIN_EDGE to MAX_EDGE, evenly spaced in the db domain
+        let nr_edges: usize = self.size - 1;
+        self.edges = (0..nr_edges)
+            .map(|x| {
+                Self::db_to_linear(
+                    MIN_EDGE + x as f32 * ((MAX_EDGE - MIN_EDGE) / (nr_edges as f32 - 1.0)),
+                )
+            })
+            .collect::<Vec<_>>()
+            .try_into()
+            .unwrap();
+
+        self.decay_weight = Self::decay_weight(self.decay, self.sample_rate);
+    }
+
+    fn db_to_linear(db: f32) -> f32 {
+        10.0_f32.powf(db / 20.0)
+    }
+
+    // Function to find the bin for a given linear audio value
+    fn find_bin(&self, value: f32) -> usize {
+        // Check if the value is smaller than the first edge
+        if value < self.edges[0] {
+            // if value < f32::EPSILON {
+            // if value == 0.0 {
+            return 0;
+        }
+        // Check if the value is larger than the last edge
+        if value > *self.edges.last().unwrap() {
+            return self.edges.len() as usize;
+        }
+        // Binary search to find the bin for the given value
+        let mut left = 0;
+        let mut right = self.edges.len() - 1;
+
+        while left <= right {
+            let mid = left + (right - left) / 2;
+            if value >= self.edges[mid] {
+                left = mid + 1;
+            } else {
+                right = mid - 1;
+            }
+        }
+        // Return the bin index
+        left as usize
+    }
+}
+
+impl VisualizerBuffer<f32> for HistogramBuffer {
+    fn enqueue(self: &mut Self, value: f32) {
+        let value = value.abs();
+        // don't enqueue silence
+        if value > 0.0 {
+            let bin_index = self.find_bin(value);
+            self.data[bin_index] += (1.0 - self.decay_weight); // Increment the count for the bin
+            for i in 0..self.size - 1 {
+                // decay all values
+                self.data[i] *= self.decay_weight;
+            }
```

Comment:
This is kind of a nitpick, but iirc, you usually increment a single bin *after* decrementing all bins. So maybe swap those steps :)